Lineage is not getting tracked for subquery IN

Lineage is not getting tracked for jobs where IN clause is involved with left hand side of it being not null and when right side of expression yields null. 

This was not present earlier as the bug was recently identified in spark and back ported to certain versions of spark which is causing the Spline parser to fail. 

More details 
Spakr PR https://github.com/apache/spark/pull/41094 

Spark Issue https://issues.apache.org/jira/browse/SPARK-43413

Spline is throwing below error after above changes in Spark. Below has been added recently. By default, nullable is returned false, hence need to handle this case specifically in Spline.

[https://github.com/apache/spark/commit/2e56821830019765bf8530e0e6a8a5abd6125293]


https://github.com/apache/spark/blob/2e56821830019765bf8530e0e6a8a5abd6125293/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/subquery.scala

```
java.lang.AssertionError: assertion failed: ListQuery nullability is not defined
	at scala.Predef$.assert(Predef.scala:223)
	at org.apache.spark.sql.catalyst.expressions.ListQuery.nullable(subquery.scala:475)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convertDataType(ExpressionConverter.scala:106)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:97)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.za$co$absa$commons$lang$CachingConverter$$super$convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47)
	at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:209)
	at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:206)
	at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:82)
	at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47)
	at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.converter.ExprToRefConverter.convert(ExprToRefConverter.scala:42)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.$anonfun$convertChildren$1(ExpressionConverter.scala:111)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convertChildren(ExpressionConverter.scala:111)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:98)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.za$co$absa$commons$lang$CachingConverter$$super$convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47)
	at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:209)
	at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:206)
	at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:82)
	at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47)
	at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.converter.ExprToRefConverter.convert(ExprToRefConverter.scala:42)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.$anonfun$convertChildren$1(ExpressionConverter.scala:111)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convertChildren(ExpressionConverter.scala:111)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:98)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.za$co$absa$commons$lang$CachingConverter$$super$convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47)
	at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:209)
	at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:206)
	at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:82)
	at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47)
	at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.converter.ExprToRefConverter.convert(ExprToRefConverter.scala:42)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.$anonfun$convertChildren$1(ExpressionConverter.scala:111)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.Iterator.foreach(Iterator.scala:941)
	at scala.collection.Iterator.foreach$(Iterator.scala:941)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1429)
	at scala.collection.IterableLike.foreach(IterableLike.scala:74)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:73)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:56)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.AbstractTraversable.map(Traversable.scala:108)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convertChildren(ExpressionConverter.scala:111)
	at za.co.absa.spline.harvester.converter.ExpressionConverter.convert(ExpressionConverter.scala:57)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.za$co$absa$commons$lang$CachingConverter$$super$convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.commons.lang.CachingConverter.$anonfun$convert$1(converters.scala:47)
	at scala.collection.mutable.MapLike.getOrElseUpdate(MapLike.scala:209)
	at scala.collection.mutable.MapLike.getOrElseUpdate$(MapLike.scala:206)
	at scala.collection.mutable.AbstractMap.getOrElseUpdate(Map.scala:82)
	at za.co.absa.commons.lang.CachingConverter.convert(converters.scala:47)
	at za.co.absa.commons.lang.CachingConverter.convert$(converters.scala:44)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.builder.plan.PlanOperationNodeBuilder$$anon$2.convert(PlanOperationNodeBuilder.scala:46)
	at za.co.absa.spline.harvester.converter.ExprToRefConverter.convert(ExprToRefConverter.scala:42)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$$nestedInanonfun$valueDecomposer$1$1.applyOrElse(OperationParamsConverter.scala:42)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter$$anonfun$$nestedInanonfun$valueDecomposer$1$1.applyOrElse(OperationParamsConverter.scala:35)
	at scala.PartialFunction$OrElse.apply(PartialFunction.scala:172)
	at za.co.absa.spline.harvester.converter.ValueDecomposer.$anonfun$recursion$1(ValueDecomposer.scala:41)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter.$anonfun$convert$6(OperationParamsConverter.scala:59)
	at scala.collection.TraversableLike$WithFilter.$anonfun$map$2(TraversableLike.scala:827)
	at scala.collection.immutable.Map$Map2.foreach(Map.scala:159)
	at scala.collection.TraversableLike$WithFilter.map(TraversableLike.scala:826)
	at za.co.absa.spline.harvester.converter.OperationParamsConverter.convert(OperationParamsConverter.scala:54)
	at za.co.absa.spline.harvester.builder.plan.GenericPlanNodeBuilder.build(GenericPlanNodeBuilder.scala:41)
	at za.co.absa.spline.harvester.builder.plan.GenericPlanNodeBuilder.build(GenericPlanNodeBuilder.scala:25)
	at za.co.absa.spline.harvester.LineageHarvester.$anonfun$harvest$6(LineageHarvester.scala:68)
	at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:238)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.collection.TraversableLike.map(TraversableLike.scala:238)
	at scala.collection.TraversableLike.map$(TraversableLike.scala:231)
	at scala.collection.immutable.List.map(List.scala:298)
	at za.co.absa.spline.harvester.LineageHarvester.$anonfun$harvest$4(LineageHarvester.scala:68)
	at scala.Option.flatMap(Option.scala:271)
	at za.co.absa.spline.harvester.LineageHarvester.harvest(LineageHarvester.scala:61)
	at za.co.absa.spline.agent.SplineAgent$$anon$1.$anonfun$handle$1(SplineAgent.scala:91)
	at za.co.absa.spline.agent.SplineAgent$$anon$1.withErrorHandling(SplineAgent.scala:100)
	at za.co.absa.spline.agent.SplineAgent$$anon$1.handle(SplineAgent.scala:72)
	at za.co.absa.spline.harvester.listener.QueryExecutionListenerDelegate.onSuccess(QueryExecutionListenerDelegate.scala:28)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1(SplineQueryExecutionListener.scala:41)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.$anonfun$onSuccess$1$adapted(SplineQueryExecutionListener.scala:41)
	at scala.Option.foreach(Option.scala:407)
	at za.co.absa.spline.harvester.listener.SplineQueryExecutionListener.onSuccess(SplineQueryExecutionListener.scala:41)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:155)
	at org.apache.spark.sql.util.ExecutionListenerBus.doPostEvent(QueryExecutionListener.scala:131)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:118)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:102)
	at org.apache.spark.sql.util.ExecutionListenerBus.postToAll(QueryExecutionListener.scala:131)
	at org.apache.spark.sql.util.ExecutionListenerBus.onOtherEvent(QueryExecutionListener.scala:135)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent(SparkListenerBus.scala:102)
	at org.apache.spark.scheduler.SparkListenerBus.doPostEvent$(SparkListenerBus.scala:28)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.scheduler.AsyncEventQueue.doPostEvent(AsyncEventQueue.scala:37)
	at org.apache.spark.util.ListenerBus.postToAll(ListenerBus.scala:118)
	at org.apache.spark.util.ListenerBus.postToAll$(ListenerBus.scala:102)
	at org.apache.spark.scheduler.AsyncEventQueue.super$postToAll(AsyncEventQueue.scala:105)
	at org.apache.spark.scheduler.AsyncEventQueue.$anonfun$dispatch$1(AsyncEventQueue.scala:105)
	at scala.runtime.java8.JFunction0$mcJ$sp.apply(JFunction0$mcJ$sp.java:23)
	at scala.util.DynamicVariable.withValue(DynamicVariable.scala:62)
	at org.apache.spark.scheduler.AsyncEventQueue.org$apache$spark$scheduler$AsyncEventQueue$$dispatch(AsyncEventQueue.scala:100)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.$anonfun$run$1(AsyncEventQueue.scala:96)
	at org.apache.spark.util.Utils$.tryOrStopSparkContext(Utils.scala:1599)
	at org.apache.spark.scheduler.AsyncEventQueue$$anon$2.run(AsyncEventQueue.scala:96)
```

--- a/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
@@ -79,6 +79,16 @@ class ExpressionConverter(
         name = e.prettyName,
         params = getExpressionParameters(e)
       )
+    
+    case e: sparkExprssions.ListQuery =>
+      FunctionalExpression(
+        id = idGen.nextId(),
+        dataType = convertDataTypeNonNullable(e),
+        childRefs = Seq.empty,
+        extra = createExtra(e, ExprV1.Types.GenericLeaf),
+        name = e.prettyName,
+        params = getExpressionParameters(e)
+      )
 
     case _: sparkExprssions.WindowSpecDefinition
          | _: sparkExprssions.WindowFrame =>
@@ -104,6 +114,11 @@ class ExpressionConverter(
 
   private def convertDataType(expr: sparkExprssions.Expression) = {
     val dataType = dataTypeConverter.convert(expr.dataType, expr.nullable)
+    Some(dataType.id)
+  }
+
+  private def convertDataTypeNonNullable(expr: sparkExprssions.Expression) = {
+    val dataType = dataTypeConverter.convert(expr.dataType, false)
     Some(dataType.id)
   }
 

--- a/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
@@ -84,7 +84,7 @@ class ExpressionConverter(
       FunctionalExpression(
         id = idGen.nextId(),
         dataType = convertDataTypeNonNullable(e), //refer issue #700 for details why the standard method is not used
-        childRefs = Seq.empty,
+        childRefs = convertChildren(e),
         extra = createExtra(e, ExprV1.Types.GenericLeaf),
         name = e.prettyName,
         params = getExpressionParameters(e)

--- a/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
@@ -118,7 +118,7 @@ class ExpressionConverter(
   }
 
   private def convertDataTypeNonNullable(expr: sparkExprssions.Expression) = {
-    val dataType = dataTypeConverter.convert(expr.dataType, false)
+    val dataType = dataTypeConverter.convert(expr.dataType, nullable = false)
     Some(dataType.id)
   }
 

--- a/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
@@ -83,7 +83,7 @@ class ExpressionConverter(
     case e: sparkExprssions.ListQuery =>
       FunctionalExpression(
         id = idGen.nextId(),
-        dataType = convertDataTypeNonNullable(e),
+        dataType = convertDataTypeNonNullable(e), //refer issue #700 for details why the standard method is not used
         childRefs = Seq.empty,
         extra = createExtra(e, ExprV1.Types.GenericLeaf),
         name = e.prettyName,

--- a/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
+++ b/core/src/main/scala/za/co/absa/spline/harvester/converter/ExpressionConverter.scala
@@ -85,7 +85,7 @@ class ExpressionConverter(
         id = idGen.nextId(),
         dataType = convertDataTypeNonNullable(e), //refer issue #700 for details why the standard method is not used
         childRefs = convertChildren(e),
-        extra = createExtra(e, ExprV1.Types.GenericLeaf),
+        extra = createExtra(e, ExprV1.Types.Generic),
         name = e.prettyName,
         params = getExpressionParameters(e)
       )

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -206,6 +206,7 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
 
   it should "return dataTypeId of any value when a ListQuery expression is called" in {
     val expression = mock[ListQuery]
+    when(expression.dataType).thenReturn(StringType)
 
     inside(converter.convert(expression)) {
       case fe: FunctionalExpression =>

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -202,6 +202,14 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
         fe.params shouldNot contain key "otherExpression"
     }
   }
+  it should "return dataTypeId of any value when a ListQuery expression is called" in {
+    val expression = mock(classOf[ListQuery])
+
+    inside(converter.convert(expression)) {
+      case fe: FunctionalExpression =>
+        (fe.dataTypeId.isEmpty || fe.dataTypeId.isDefined) should be(true)
+    }
+  }
 }
 
 object ExpressionConverterSpec {

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -206,8 +206,10 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
 
   it should "throw exception when nullable property is accessed for ListQuery else should return a dataType" in {
     val expression = mock[ListQuery]
+    val aChild = Literal("this is a child")
     when(expression.dataType).thenReturn(StringType)
     when(expression.nullable).thenThrow(new RuntimeException("accessing nullable property not supported for ListQuery"))  // should not be called
+    when(expression.children).thenReturn(Seq(aChild))
     inside(converter.convert(expression)) {
       case fe: FunctionalExpression =>
         fe.dataType.isDefined should be(true)
@@ -248,8 +250,33 @@ object ExpressionConverterSpec {
     protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression = copy(children = newChildren)
   }
 
+
+  case class FooListQuery(
+    returnType: DataType,
+    nullable: Boolean,
+    children: Seq[Expression],
+    any: Any
+  ) extends ListQuery {
+
+    def this(aParamOfSecondaryConstructor: String) = this(NullType, true, Nil, null, null, null, -1, -1, null, None, null, null)
+
+    val additionalPropertyNotFromConstructor = "this should not be captured"
+
+    override def dataType: DataType = returnType
+
+    override def eval(input: InternalRow): Any = ()
+
+    override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = null
+
+    protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression = copy(children = newChildren)
+  }
+
   object Foo {
     val empty = new Foo("does not matter")
+  }
+
+  object FooListQuery {
+    val empty = new FooListQuery("testing list query expression")
   }
 
   object Bar {

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -19,6 +19,7 @@ package za.co.absa.spline.harvester.converter
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions.codegen.{CodegenContext, ExprCode}
 import org.apache.spark.sql.catalyst.expressions.{CaseWhen, Expression, Literal}
+import org.apache.spark.sql.catalyst.expressions.{CaseWhen, Expression, Literal,ListQuery}
 import org.apache.spark.sql.types.DataTypes.NullType
 import org.apache.spark.sql.types._
 import org.mockito.Mockito._
@@ -202,12 +203,13 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
         fe.params shouldNot contain key "otherExpression"
     }
   }
+
   it should "return dataTypeId of any value when a ListQuery expression is called" in {
-    val expression = mock(classOf[ListQuery])
+    val expression = mock[ListQuery]
 
     inside(converter.convert(expression)) {
       case fe: FunctionalExpression =>
-        (fe.dataTypeId.isEmpty || fe.dataTypeId.isDefined) should be(true)
+        (fe.dataType.isEmpty || fe.dataType.isDefined) should be(true)
     }
   }
 }

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -204,13 +204,13 @@ class ExpressionConverterSpec extends AnyFlatSpec with OneInstancePerTest with M
     }
   }
 
-  it should "return dataTypeId of any value when a ListQuery expression is called" in {
+  it should "throw exception when nullable property is accessed for ListQuery else should return a dataType" in {
     val expression = mock[ListQuery]
     when(expression.dataType).thenReturn(StringType)
-
+    when(expression.nullable).thenThrow(new RuntimeException("accessing nullable property not supported for ListQuery"))  // should not be called
     inside(converter.convert(expression)) {
       case fe: FunctionalExpression =>
-        (fe.dataType.isEmpty || fe.dataType.isDefined) should be(true)
+        fe.dataType.isDefined should be(true)
     }
   }
 }

--- a/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
+++ b/core/src/test/scala/za/co/absa/spline/harvester/converter/ExpressionConverterSpec.scala
@@ -250,33 +250,8 @@ object ExpressionConverterSpec {
     protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression = copy(children = newChildren)
   }
 
-
-  case class FooListQuery(
-    returnType: DataType,
-    nullable: Boolean,
-    children: Seq[Expression],
-    any: Any
-  ) extends ListQuery {
-
-    def this(aParamOfSecondaryConstructor: String) = this(NullType, true, Nil, null, null, null, -1, -1, null, None, null, null)
-
-    val additionalPropertyNotFromConstructor = "this should not be captured"
-
-    override def dataType: DataType = returnType
-
-    override def eval(input: InternalRow): Any = ()
-
-    override def doGenCode(ctx: CodegenContext, ev: ExprCode): ExprCode = null
-
-    protected def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression = copy(children = newChildren)
-  }
-
   object Foo {
     val empty = new Foo("does not matter")
-  }
-
-  object FooListQuery {
-    val empty = new FooListQuery("testing list query expression")
   }
 
   object Bar {

